Refactor bug reporting into plugins / Remove bug report mako

hexylena opened the pull request:
I've extended bug reports to use a more plugin style infrastructure. Instead of "user submitted = email, tool error = sentry (if configured)", the admin can decide how they want to handle bug reports. Do they want to submit sentry reports only when the user clicks the bug report? Do they want to send emails on every single failed tool? Do they have some third party system they wish to integrate with?

I've refactored Email + Sentry into two plugins, and provided a 'JSON' output as a reference plugin for more implementations.  

As a result, the post-submission screen has been extended to show multiple messages. One can imagine a third party bug reporting plugin informing the user "Your bug reference number is [#123]()"

![utvalg_240](https://user-images.githubusercontent.com/458683/28105634-d484db9c-66cf-11e7-93fc-b59ae2031977.png)

And the output of the example JSON plugin:
```json
{
  "traceback": null, 
  "stdout": "", 
  "command_line": "echo 'Fail' > /home/hxr/work/galaxy/database/files/000/dataset_94.dat; exit 1;", 
  "user": {
    "username": "test", 
    "deleted": false, 
    "email": "test@domain.edu", 
    "last_password_change": "2017-06-02T19:18:11.278864", 
    "active": false, 
    "model_class": "User", 
    "id": 1
  }, 
  "message": "Testing a bug report", 
  "id": 94, 
  "info": null, 
  "tool_version": "1.0", 
  "tool_xml": "./tools/error.xml", 
  "exit_code": 1, 
  "handler": "main", 
  "stderr": "Fatal error: Exit code 1 ()\n", 
  "email": "hxr@localhost"
}
```


- cc @moffmade, who I believe wanted to implement a "github" plugin.
- Thanks @jmchilton for a great example of the plugin infrastructure with the metrics infrastructure.
- It sure is nice being between jobs and having time for fun stuff.

hexylena committed:
Refactor bug reporting into plugins

hexylena committed:
flake8

dannon commented:
This is a great idea.

Can we use a simple dictionary endpoint and clientside route, similar to @guerler's recent work swapping out grids, to avoid backsliding a bit by adding a new mako?  I'd be happy to help with that prior to merge it'd be useful.

And (this one's really not important, just curious!) I'm not super keen on the 'sink' nomenclature.  Did you consider any other names?

hexylena commented:
> Can we use a simple dictionary endpoint and clientside route, similar to @guerler's recent work swapping out grids, to avoid backsliding a bit by adding a new mako?

We sure could. I have no experience converting the mako to JS. I hope to practice with the job info page at some point soon, maybe thursday or friday (once I'm done packing my house). If I promise to do it before the next release, would you be ok if I refactor this in a separate PR?

> And (this one's really not important, just curious!) I'm not super keen on the 'sink' nomenclature. Did you consider any other names?

I hadn't, actually. Source/Sink was the first thing that came to mind. Do you have something you'd prefer?

hexylena committed:
flake8 import

bgruening commented:
@anuprulez could help migrating this. He is an expert in this kind of stuff after it is merged, if this helps.
Great feature @erasche!

bgruening commented:
@erasche can I bribe you to add a slide to the admin training after this is merged :)

hexylena commented:
@bgruening was hoping to learn how to do client end stuff with this / job info page, before @anuprulez can get to them ;) and yes, of course, should be documented. that should be a blocking problem with this PR, i'll fix it now.

hexylena commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1809,27 +1809,11 @@ def requires_setting_metadata( self ):
-                }
-            )
+        for dataset in job.output_datasets:
+            self.app.error_reports.default_error_sink.submit_report(dataset, job, tool, user_submission=False)
```
@jmchilton is this the right way to do this? I get the impression that it isn't, but I wasn't understanding the lambda expression that I copied from your metrics. I.e. https://github.com/galaxyproject/galaxy/pull/4305/files#diff-218b8eec0cd1aaab08f26181429f4ba0R19

jmchilton replied:
The job instrumenters you copied this from are per-job-destination. So different job destinations can have different job metrics collected. The default applies to any that aren't specified on a per destination basis - which is likely every destination ever since this wasn't really documented. I'd probably argue the error reporting should be overridable on a per destination basis for the same reason - but since I didn't document it replicating the full complexity of job metrics is not something I'd really require from you.

But to answer the question about the odd line:

```
    self.error_sinks = collections.defaultdict(lambda: self.default_error_sink)
```

It is meant to be used like:

```
    self.app.error_reports.error_sinks[job.destination_id].submit_report(dataset, job, tool, user_submission=False)
```

In this case, the lambda is just telling the ``defaultdict`` which set of error sink plugins to use if no per-destination one has been explicitly set. I'd make that change just to bring it inline with job metrics and because you have the job available right there - it allows us to grow in that direction in the future (per-destination error reporting) - even if you don't add the job conf hooks now.

Ideally everything is overridable at the destination this way. The nature of job destinations is also such that this means it is overridable from dynamic job destinations and on a per-tool basis for instance also.

hexylena committed:
document

hexylena committed:
Allow sentry to be verbose

hexylena commented:
![utvalg_241](https://user-images.githubusercontent.com/458683/28129798-3d1a7032-6724-11e7-86ef-870b45183fa8.png)

sentry plugin can now be verbose. (I.e. display a message to the user when it is configured to be run on user-submission).

dannon commented:
Generally speaking, I think we should for new features include new style client endpoints in the initial merge, unless there's pressure to get something in.  It doesn't hurt anything for PRs to stay open and continue getting worked on, and we more often 'lose' issues and TODOs than we do open PRs.  That said, I'm definitely not a `-` on this until it's done, I'd just prefer it the other way 'round.

And, maybe just 'Plugin'?  Sink just seems overly technical (if completely correct) and harder to grok at a glance, to me putting myself in the shoes of a naive browser of the codebase.

hexylena commented:
@dannon all fair. We have a ways to go until 17.09, so I'll work on a client side replacement as well, though that will significantly delay this PR being ready.

I had originally called it ErrorReporter, but I couldn't easily visually distinguish it from "ErrorReports" (i.e. https://github.com/galaxyproject/galaxy/pull/4305/files#diff-218b8eec0cd1aaab08f26181429f4ba0R12) so changed it to sink. ErrorPlugin will be fine, I can do this as well.

dannon commented:
@erasche Awesome, thanks for being understanding, and let me know if I can help out at all!  This is great stuff, and a valuable improvement.

jmchilton commented:
@erasche If you just change that config file extension to .yml - none of your other code would need to change (e.g. this https://github.com/common-workflow-language/cwltool/commit/b19f60d9b430852b4cb873ac020b72e67f78e2a9). The plugin framework should just work after you update the default config paths. I made the new workflow scheduler plugins file YAML by default - it is the direction I'd like to head. Are you open to swapping this out - or would you prefer to merge as is?

For name suffix, I prefer Sink to Plugin but am okay with either. Either ErrorCollector or ErrorRespnder probably would have been the name I would have used - but that is hardly an endorsement.

dannon commented:
Maybe 'handlers' instead of 'sinks' or 'plugins'?  So, ErrorHandler, EmailErrorHandler, etc. 

edit:  (just to reiterate, this is not an important issue for me, just struck me as awkwardly named and I was curious -- do whatever seems right to you)

hexylena commented:
(Last day of packing, will respond tonight. Thanks for the great feedback y'all!)

jmchilton commented on lib/galaxy/config.py:
```diff
@@ -43,6 +43,7 @@
     job_config_file=['config/job_conf.xml', 'job_conf.xml'],
     tool_destinations_config_file=['config/tool_destinations.yml', 'config/tool_destinations.yml.sample'],
     job_metrics_config_file=['config/job_metrics_conf.xml', 'job_metrics_conf.xml', 'config/job_metrics_conf.xml.sample'],
+    error_report_file=['config/error_report.xml', 'error_report.xml', 'config/error_report.xml.sample'],
```
I'd remove this middle entry - we allow them to sit in ``$GALAXY_ROOT`` for some older configs for legacy reasons - they were there before we introduced the ``config/`` directory - we don't want new config files in ``$GALAXY_ROOT``.

hexylena committed:
Per @jmchilton comment, rewrite as yml

hexylena committed:
add bug reporter

hexylena committed:
redirect to new url

hexylena committed:
add clientside route

hexylena committed:
move css rule to a more generic location

hexylena committed:
change report code

hexylena committed:
refactor into an API

hexylena committed:
Handle error report submission

hexylena committed:
refactor a bit

hexylena committed:
Add extra parameter to allow multiple messages

hexylena committed:
Change invocation to pass user information

hexylena committed:
Link in expected places

hexylena committed:
Switch biostars to using HTML submission

hexylena committed:
Now unused

hexylena committed:
Requested renames, sink + yml

hexylena committed:
more renames

hexylena committed:
forgot static files

hexylena commented:
Per request, it was made into an API / client side thing. Reused style from the job info page for visual distinctiveness

![auswahl_002](https://user-images.githubusercontent.com/458683/29030380-11ed3a72-7b8b-11e7-8f9c-5f30c6872ddd.png)

I've integrated the BioStars reporting as well, tested, it works.

![auswahl_003](https://user-images.githubusercontent.com/458683/29030381-11ef0316-7b8b-11e7-812b-d921832ba28f.png)

As part of that I switched the biostars submission over to use the relatively new HTML error report template since it looks *so* much nicer.

![auswahl_004](https://user-images.githubusercontent.com/458683/29030382-11f1959a-7b8b-11e7-8c7e-614b7d7862e2.png)

hexylena committed:
merge

hexylena committed:
another merge

hexylena commented:
This is ready for review :) Rebased against dev.

hexylena committed:
address flaking

hexylena committed:
remove comment

hexylena commented:
Not sure why these tests are failing. @martenson?

mvdbeek commented:
it errors out with:
```
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/webapps/galaxy/buildapp.py", line 58, in paste_app_factory
    app = galaxy.app.UniverseApplication( global_conf=global_conf, **kwargs )
  File "/Users/mvandenb/src/galaxy/lib/galaxy/app.py", line 107, in __init__
    self.error_reports = ErrorReports( self.config.error_report_file, app=self )
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/error_reports/__init__.py", line 18, in __init__
    self.default_error_plugin = ErrorPlugin.from_file(self.plugin_classes, conf_file, **kwargs)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/error_reports/__init__.py", line 73, in from_file
    plugins_source = plugin_config.plugin_source_from_path(conf_file)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/util/plugin_config.py", line 78, in plugin_source_from_path
    return ('xml', ElementTree.parse( path ).getroot())
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/xml/etree/ElementTree.py", line 1182, in parse
    tree.parse(source, parser)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/xml/etree/ElementTree.py", line 656, in parse
    parser.feed(data)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/xml/etree/ElementTree.py", line 1653, in feed
    self._raiseerror(v)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/xml/etree/ElementTree.py", line 1517, in _raiseerror
    raise err
ParseError: not well-formed (invalid token): line 1, column 1
```

hexylena commented:
<s>ahhh interesting. @jmchilton you said switching to yml would be fine, did I miss something when I swapped it out? The relevant commits would be https://github.com/galaxyproject/galaxy/pull/4305/commits/fafdd6be0c7bb70fad8708987cb1f507aca994fa and https://github.com/galaxyproject/galaxy/pull/4305/commits/f445e60b6dacaa7fb228949426cefad8d0453d4f</s>

hexylena commented:
Oh, no, it is failing consistently for me now, ok, give me a minute. (Got home, different machine)

hexylena committed:
Also load yaml sample files. cc @jmchilton, is this ok?

hexylena committed:
correct name to type

jmchilton commented:
@erasche  a3cc4c5 looks like a good fix to me - thanks for that!

hexylena committed:
fix import order

jmchilton committed:
Remove unneeded messages.mako.

jmchilton commented:
- Playing around with this - it looks like sentry and biostars should still just work under the new default configuration - people shouldn't have to update anything in their configs when upgrading Galaxy to re-enable this I think. That is good.
- It looks like you still get an error box to write an e-mail even if Galaxy knows it isn't going to work. This is a pre-existing problem though - that doesn't need to be fixed with this PR. The error message was a little nicer before - I got a nice red box and now it is just text on a black background.  This isn't a high priority for me - I'm happy to merge without that fixed but if we could make that box red and make it clear there was a problem like before that would be great?  <img width="1415" alt="screen shot 2017-08-10 at 10 34 34 am" src="https://user-images.githubusercontent.com/216771/29175659-a05aeb14-7db7-11e7-9fa4-ea5f4d84e552.png">
- The one other thing I would like to see though is the outdated code axed. It looks like messages.mako is added and not needed anymore - as well as the helper that called it. I think this commit https://github.com/jmchilton/galaxy/commit/d9a049292be77ca22d948b66212dac0c57b0d2a9 should be enough if you want to cherry-pick it.

hexylena committed:
fix css class

hexylena committed:
not a useful test in this location

hexylena committed:
rebuild client

jmchilton commented:
Thanks @erasche - this is really cool stuff!

hexylena commented:
> people shouldn't have to update anything in their configs when upgrading Galaxy to re-enable this I think. That is good.

I tried to be very careful about this, if I have missed anything I want to be sure to catch it now. This will be a nice feature, but as you've mentioned, I don't want to negatively affect anyone with these changes or cause people to lose bug reports.

> It looks like you still get an error box to write an e-mail even if Galaxy knows it isn't going to work. This is a pre-existing problem though - that doesn't need to be fixed with this PR.

The error box's contents will also be included in any other plugin that wants to make use of it. E.g. the sentry plugin will accept a message. So in this way, I think that it is fine for the box to be essentially a permanent fixture. In the email-only world, I agree that it was weird to show the box, knowing ahead of time that it would break.

The only similar case here was if the admin actively disables all error plugins. We could probably stand to show a message if there were no enabled backends for error reporting, but I think that that is a corner case.

> The error message was a little nicer before - I got a nice red box and now it is just text on a black background.

The class names changed when I changed the css a bit, and it looks like I forgot to update the email plugin. Good catch, thanks.

> The one other thing I would like to see though is the outdated code axed. It looks like messages.mako is added and not needed anymore - as well as the helper that called it. I think this commit jmchilton/galaxy@d9a0492 should be enough if you want to cherry-pick it.

Oooh, awesome, thanks for doing that. Added.

hexylena commented:
apparently I never hit submit, sorry about that :) thanks for the merge.
